src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -0,0 +1,273 @@
+import { html, LitElement } from "@polymer/lit-element";
+import { repeat } from "lit-html/directives/repeat";
+import { TemplateResult } from "lit-html";
+import "@polymer/paper-checkbox/paper-checkbox";
+import "@polymer/paper-icon-button/paper-icon-button";
+import "@polymer/paper-input/paper-input";
+import "@polymer/paper-item/paper-icon-item";
+import "@polymer/paper-item/paper-item-body";
+import "@polymer/paper-button/paper-button";
+
+import "../../../components/ha-card";
+import "../../../components/ha-icon";
+
+import { hassLocalizeLitMixin } from "../../../mixins/lit-localize-mixin";
+import { HomeAssistant } from "../../../types";
+import { LovelaceCard, LovelaceConfig } from "../types";
+import {
+  clearItems,
+  fetchItems,
+  addItem,
+  completeItem,
+  saveEdit,
+  ShoppingListItem,
+} from "../../../data/shopping-list";
+
+interface Config extends LovelaceConfig {
+  title?: string;
+}
+
+class HuiShoppingListCard extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  public hass?: HomeAssistant;
+  private _config?: Config;
+  private _items?: ShoppingListItem[];
+  private _unsubEvents?: () => Promise<void>;
+
+  static get properties() {
+    return {
+      _config: {},
+      _items: {},
+    };
+  }
+
+  public getCardSize(): number {
+    return (
+      (this._config ? (this._config.title ? 1 : 0) : 0) +
+      (this._items ? this._items.length : 3)
+    );
+  }
+
+  public setConfig(config: Config): void {
+    this._config = config;
+    this._items = [];
+    this._fetchData();
+  }
+
+  public connectedCallback(): void {
+    super.connectedCallback();
+
+    if (this.hass) {
```

Comment:
So since you have this check, it is important that in case we get connected to the DOM and then get `hass`, we recover from that?